add http head based digest comparison to avoid dockerhub rate limits

simonaronsson opened the pull request:
We currently experience a pretty urgent issue where each check by watchtower counts towards the DockerHub rate limit. What we've managed to understand so far after discussions with the docker OSS team and others:

- DockerHub does not count HEAD requests towards the rate limit
- Docker Engine was thought to do HEAD requests where possible. This, however, turned out to not be the case for various reasons outlined in https://github.com/moby/moby/pull/41607.
- Getting this fix out to the community will take time. My best bet is that even if it's released in 20.xx, it will take years until it reaches a significant portion of our users.
- The docker registry specification is fairly detailed in how registries should behave in terms of authentication, tokens, challenges, and what request headers do what.
- GHCR and DockerHub behave almost the same, which is a relief as they, together with quay, likely make up a significant part of all registry checks done by Watchtower.
- Depending on `skopeo` or any other repo in the `containers` org. is currently not an option. It has the same issue with fetching manifest list digests as docker engine, which would lead to digest mismatches for all multi-arch tags. As a multi-arch tag leads to at least 2 GET requests toward the rate limit, this in itself makes it a deal-breaker for us.
- I've not yet been able to verify whether HEAD requests bump the pull count. Hopefully, they don't, solving some other long-lingering issues (see #429, #201).
- To make sure we don't break anything for users on older/incompatible registries, we'll make sure to fallback to GET requests if a registry doesn't behave as expected.
- The Docker OSS team has kindly offered to work with us to see if there is anything to be done, in any end, to make the user experience smoother and less resource-consuming going forward. 

I'm fairly certain I've missed something, so we'll have to go through this again before releasing, but at least we're getting somewhere. 😅

simonaronsson committed:
add http head based digest comparison to avoid dockerhub rate limits

piksel commented on pkg/registry/digest/digest_test.go:
```diff
@@ -0,0 +1,64 @@
+	DH_USERNAME = os.Getenv("CI_INTEGRATION_TEST_REGISTRY_DH_USERNAME")
+	DH_PASSWORD = os.Getenv("CI_INTEGRATION_TEST_REGISTRY_DH_PASSWORD")
+	GH_USERNAME = os.Getenv("CI_INTEGRATION_TEST_REGISTRY_DH_USERNAME")
+	GH_PASSWORD = os.Getenv("CI_INTEGRATION_TEST_REGISTRY_DH_PASSWORD")
```
```suggestion
	GH_USERNAME = os.Getenv("CI_INTEGRATION_TEST_REGISTRY_GH_USERNAME")
	GH_PASSWORD = os.Getenv("CI_INTEGRATION_TEST_REGISTRY_GH_PASSWORD")
```

simonaronsson replied:
Some refactoring inc. tomorrow. That change is in there, sort of! 😅

piksel replied:
Yeah, no worries. My typo alarm just went off 😁
Felt like a typical problem one could spend way too long trying to figure out why the tests fail for no good reason.

simonaronsson committed:
fix some tests, split up and refactor

some wonky regression introduced by docker dependencies when running on darwin. see https://github.com/ory/dockertest/issues/212 for more info.
will have a look at this next

simonaronsson committed:
fix linter issues

simonaronsson committed:
build around container/image, as they depend on a package currently broken on macos

simonaronsson committed:
fix regression and cleanup tests

simonaronsson committed:
increase test coverage and fix some minor bugs

simonaronsson committed:
add some more tests and debug logs

simonaronsson committed:
workaround x/sys issue on windows

simonaronsson committed:
fix linting, formatting. tidy up go mod

piksel commented on go.mod:
```diff
@@ -53,7 +56,8 @@ require (
 	github.com/zmap/zlint v1.0.2 // indirect
-	golang.org/x/net v0.0.0-20190522155817-f3200d17e092
+	golang.org/x/net v0.0.0-20200520004742-59133d7f0dd7
+	golang.org/x/sys v0.0.0-20201117170446-d9b008d0a637 // indirect
```
```suggestion
	golang.org/x/net v0.0.0-20191004110552-13f9640d40b9
	golang.org/x/sys v0.0.0-20190830141801-acfa387b8d69 // indirect
```
also works, and should lead to fewer side effects?
the only downside is ginko and gomega that cannot be upgraded past:
```
	github.com/onsi/ginkgo v1.11.0
	github.com/onsi/gomega v1.10.0
```
Gist of full go.mod: https://gist.github.com/piksel/03e87dcb7d4166ce5ab8a20ccf93aa9a

simonaronsson replied:
What side effects are you thinking of?

piksel replied:
The unknown ones :) The override makes it essentially just treat it as if it were another version, which it may not work correctly with.

simonaronsson replied:
Fixed.

piksel replied:
Looking at the specification, it seems like the override SHOULD work the same way as my solution. Why it insists on upgrading certain packages is beyond me.

simonaronsson committed:
fix tests, simplify and integrate credentials properly

simonaronsson committed:
fix accept headers and switch default interval to 24h

piksel commented on pkg/registry/auth/auth.go:
```diff
@@ -0,0 +1,168 @@
+		key := kv[0]
+		val := strings.Trim(kv[1], "\"")
+		values[key] = val
+	}
```
The values map could be initialized with `len(pairs)`

simonaronsson committed:
initialize map using length of pair array

piksel commented on pkg/registry/auth/auth.go:
```diff
@@ -0,0 +1,168 @@
+	"github.com/sirupsen/logrus"
+	"io/ioutil"
+	"net/http"
+	url2 "net/url"
```
🤨

simonaronsson committed:
remove redundant context and ignore coverage output
